### ISSUE TITLE
Nsock consistently fails to connect to legacy SSL implementations

### ISSUE THREAD

**Issue by nnposter** [nmap/nmap#511]
There are some older TLS/SSL implementations that fail to complete the handshake with TLS 1.1/1.2 clients. The flow is roughly the same:

1. Client sends a Client Hello message with version 0x303 (TLS 1.2)
2. Server responds with a Server Hello message with version 0x301 (TLS 1.0)
3. Client obliges and sends Client Key Exchange and Change Cipher Spec messages with version 0x301
4. Server ACKs
5. In some cases Server sends an Alert (Bad Record MAC)
6. Server sends a FIN

Known offenders:
* IBM HTTP Server older than 7.0.4.25
* Dell ERA and DRAC cards
* HP iLO cards

Browsers like IE or [Firefox](https://support.mozilla.org/en-US/questions/1056008) implemented a workaround in that they attempt to reconnect with TLS 1.0. Perhaps it would be beneficial if Nsock provided a similar workaround by default.

**Comment by dmiller-nmap** [comment]
SSL connection issues are going to continue to be a problem as OpenSSL drops support for SSLv2 (default on most systems) and SSLv3 (will be completely gone in OpenSSL 1.1). We need to come up with a strategy for maintaining the broadest support possible.

In the meantime, @nnposter do you have public examples of systems that react this way so we can test potential fixes?

**Comment by nnposter** [comment]
Unfortunately I do not have any suitable Internet-facing end-point at the moment but I would be happy to run a test code.

Just to make it clear, all three listed offenders are fine when the client comes in as TLS 1.0. No need to drop all the way to SSL.

**Comment by nnposter** [comment]
Jacek probably stumbled on yet another misbehavior at instalator.upc.pl:443, reported [here](http://seclists.org/nmap-dev/2017/q1/74). The server closes the connection (with a FIN) right after the Client Hello message. It does not say what it did not like. It does not respond with Server Hello. There are no problems when connecting as a TLS 1.0 client.

Not working: `openssl s_client -connect instalator.upc.pl:443`
Working: `openssl s_client -connect instalator.upc.pl:443 -tls1`

This suggests that the proposed workaround would work in this scenario as well.

@d33tah FYI

**Comment by dmiller-nmap** [comment]
@nnposter The problem @deetah pointed out is not really a problem with Ncat as much as it is with the server. My own installation of `openssl s_client` connects fine without additional arguments. The server suffers from [Long Handshake Intolerance](https://github.com/ssllabs/research/wiki/Long-Handshake-Intolerance), so the more ciphers we support, the less likely the server is to respond. I just pushed  eb0bb25 which allows ncat users to specify cipher suites. Being more restrictive (using `HIGH:!ECDH` for instance, which eliminates a bunch of extension data in addition to the cipher suites from the handshake) with the `--ssl-ciphers` option will reduce the size of the handshake, allowing a connection.

Your original issue is still valid. I think a first step would be a command-line option to specify tls versions, something like `--tls-versions 3.0,1.0` (thanks Microsoft for the naming/versioning snafu). That would require enough reorganization of the code to enable us to have ([safe!](https://tools.ietf.org/html/rfc7507)) automatic fallback if we choose in the future.

**Comment by d33tah** [comment]
Thanks @dmiller-nmap @nnposter.

**Comment by nnposter** [comment]
You are right. I have thought that I tested for the Client Hello excess (via -cipher DES-CBC3-SHA), which failed, but I probably experienced some other glitch at that time. It is working fine now.
